### ISSUE TITLE
Error in installation 

### ISSUE THREAD

**Issue by pclucas14** [facebookresearch/vissl#198]
Hi, 

I'm trying to install the package following the  instructions 

```
conda create -n vissl python=3.8
conda activate vissl
conda install -c pytorch pytorch=1.7.1 torchvision cudatoolkit=10.2
conda install -c vissl -c iopath -c conda-forge -c pytorch -c defaults apex vissl
```

However the last step seems to be failing : I'm getting 

```
(vissl) lucaspc@devfair0224:~/repos$ conda install -c vissl -c iopath -c conda-forge -c pytorch -c default
s apex vissl

Solving environment: failed

InvalidVersionSpecError: Invalid version spec: =2.7
```

I'm not sure where the error is coming from, there may be some version conflict.

Thanks, 
Lucas

**Comment by pclucas14** [comment]
I also tried on another server, and it's still failing on the same step

```
(vissl) lpagec@dp-gpu1[pytorch]$conda install -c vissl -c iopath -c conda-forge -c pytorch -c defaults apex vissl
Collecting package metadata (current_repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.
Collecting package metadata (repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: |
Found conflicts! Looking for incompatible packages.
This can take several minutes.  Press CTRL-C to abort.
Examining conflict for olefile ca-certificates typing_extensions mkl-service Examining conflict for olefile typing_extensions mkl-service numpy ninja six Examining conflict for olefile typing_extensions mkl-service numpy ninja six Examining conflict for olefile typing_extensions mkl-service numpy ninja six Examining conflict for olefile typing_extensions mkl-service numpy ninja six Examining conflict for olefile typing_extensions mkl-service numpy ninja six Examining conflict for olefile typing_extensions mkl-service numpy ninja six Examining conflict for olefile typing_extensions mkl-service numpy ninja six Examining conflict for olefile typing_extensions mkl-service numpy ninja six Examining conflict for olefile typing_extensions mkl-service numpy freetype nExamining conflict forExamining conflict for olefile typing_extensions mkl-service numpy freetype ninja six libstdcxx-ng nump-
```

**Comment by 962086838** [comment]
Same issue on my Ubuntu 18.04 server. Here's my log. 

`$ conda install -c vissl -c iopath -c conda-forge -c pytorch -c defaults apex vissl
Collecting package metadata (current_repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.
Collecting package metadata (repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: | 
Found conflicts! Looking for incompatible packages.
This can take several minutes.  Press CTRL-C to abort.
Examining conflict for pillow mkl-service ca-certificates torchvision python typing_extensions setuptools numpy olefile numpy-base certifi ninja wheExamining conflict for pillow mkl-service ca-certificates torchvision python typing_extensions setuptools numpy olefile numpy-base certifi ninja wheExamining conflict for mkl-service torchvision pytorch wheel mkl_random python numpy-base zstd olefile ninja vissl pillow mkl_fft apex typing_extensExamining conflict for mkl-service torchvision pytorch wheel mkl_random python numpy-base zstd olefile ninja vissl pillow mkl_fft apex typing_extensExamining conflict for mkl-service torchvision libtiff pytorch wheel mkl_random python numpy-base lcms2 zstd xz olefile ninja vissl pillow mkl_fft aExamining conflict for mkl-service torchvision libtiff pytorch wheel mkl_random python numpy-base lcms2 zstd xz olefile ninja vissl pillow mkl_fft aExamining conflict for mkl-service torchvision pytorch wheel mkl_random python numpy-base olefile ninja vissl pillow mkl_fft apex typing_extensions Examining conflict for mkl-service torchvision pytorch wheel mkl_random python numpy-base olefile ninja vissl pillow mkl_fft apex typing_extensions Examining conflict for mkl-service torchvision pytorch wheel mkl_random python numpy-base readline olefile ninja vissl pillow mkl_fft apex typing_exExamining conflict for mkl-service torchvision pytorch wheel mkl_random python numpy-base readline olefile ninja vissl pillow mkl_fft apex typing_exExamining conflict for mkl-service blas mkl numpy numpy-base pytorch mkl_fft libgcc-ng mkl_random:  13%|█▋           | 6/47 [01:04<06:42,  9.82s/it]Examining conflict for mkl-service torchvision pytorch wheel mkl_random python numpy-base olefile ninja vissl pillow mkl_fft apex typing_extensions Examining conflict for mkl-service torchvision pytorch wheel mkl_random python numpy-base olefile ninja vissl pillow mkl_fft apex typing_extensions Examining conflict for mkl-service torchvision pytorch wheel mkl_random python numpy-base openssl olefile ninja vissl pillow mkl_fft apex typing_extExamining conflict for mkl-service torchvision pytorch wheel mkl_random python numpy-base openssl olefile ninja vissl pillow mkl_fft apex typing_extExamining conflict for mkl-service torchvision numpy vissl pytorch mkl_fft apex mkl_random:  21%|████               | 10/47 [01:48<06:42, 10.88s/it]Examining conflict for mkl-service torchvision libtiff pytorch wheel mkl_random zlib libpng python numpy-base lcms2 zstd freetype olefile ninja vissExamining conflict for mkl-service torchvision libtiff pytorch wheel mkl_random zlib libpng python numpy-base lcms2 zstd freetype olefile ninja vissExamining conflict for mkl-service blas torchvision typing_extensions setuptools numpy olefile numpy-base certifi ninja wheel pillow mkl_fft pip sixExamining conflict for mkl-service blas torchvision typing_extensions setuptools numpy olefile numpy-base certifi ninja wheel pillow mkl_fft pip sixExamining conflict for mkl-service torchvision pytorch wheel mkl_random python numpy-base ld_impl_linux-64 olefile ninja vissl pillow mkl_fft apex tExamining conflict for mkl-service torchvision pytorch wheel mkl_random python numpy-base ld_impl_linux-64 olefile ninja vissl pillow mkl_fft apex tExamining conflict for mkl-service torchvision mkl libtiff pytorch wheel mkl_random python blas ncurses numpy-base lcms2 zstd libffi freetype readline lz4-c olefile libedit ninja vissl pillow mkl_fft apex cudatoolkit typing_extensions setuptools numpy sqlite certifi libstdcxx-ng pip six:  30%|▎|Examining conflict for mkl-service torchvision mkl libtiff pytorch wheel mkl_random python blas ncurses numpy-base lcms2 zstd libffi freetype readline lz4-c olefile libedit ninja vissl pillow mkl_fft apex cudatoolkit typing_extensions setuptools numpy sqlite certifi libstdcxx-ng pip six:  32%|▎|Examining conflict for mkl-service torchvision pytorch wheel mkl_random python ncurses numpy-base readline olefile libedit ninja vissl pillow mkl_fft apex typing_extensions setuptools numpy sqlite certifi pip six:  32%|▎| 15/47 [02:44<05:59, 11.23s/it]                                            Examining conflict for mkl-service torchvision pytorch wheel mkl_random python ncurses numpy-base readline olefile libedit ninja vissl pillow mkl_ffExamining conflict for mkl-service blas torchvision mkl numpy numpy-base pytorch vissl mkl_fft apex mkl_random:  34%|▎| 16/47 [02:57<06:03, 11.73s/iExamining conflict for mkl-service blas torchvision mkl numpy numpy-base pytorch vissl mkl_fft apex mkl_random:  36%|▎| 17/47 [02:57<06:02, 12.08s/iExamining conflict for mkl-service torchvision pytorch wheel mkl_random python numpy-base olefile ninja vissl pillow mkl_fft apex typing_extensions Examining conflict for mkl-service torchvision pytorch wheel mkl_random python numpy-base olefile ninja vissl pillow mkl_fft apex typing_extensions Examining conflict for mkl-service torchvision numpy numpy-base wheel vissl mkl_fft six mkl_random:  38%|████▏      | 18/47 [03:21<05:43, 11.85s/it]Examining conflict for mkl-service torchvision numpy numpy-base wheel vissl mkl_fft six mkl_random:  40%|████▍      | 19/47 [03:21<05:33, 11.90s/\ ]`

**Comment by 962086838** [comment]
Same step and same issue on my Ubuntu 18.04 server.

**Comment by prigoyal** [comment]
thank you for reporting this @pclucas14  and @962086838. I am able to reproduce this too myself now which is strange as we haven't updated VISSL packages recently the uploaded packages worked seamlessly few days ago.

Assigning to @bottler to take a look at this!

**Comment by bottler** [comment]
This is an opencv weirdness.

First observation: The command works fine if I replace `python=3.8` with `python=3.7`. The version of python shouldn't matter much. This is helpful, because we have a baseline. (If you don't actually need Python3.8 and are just trying to use vissl, the rest of this comment can be ignored!)

There are several packages which exist in more than one place on anaconda cloud. I have separated out all the dependencies. The following goes straight through, with the last line only pulling in apex and vissl. When I say "straight through", I mean the conda solver doesn't seem to need to work hard.  

```
conda create -p /raid/reizenstein/jeremyconda/v4_p37 python=3.7 scikit-learn scipy parameterized
conda activate /raid/reizenstein/jeremyconda/v4_p37
conda install -c pytorch pytorch=1.7.1 torchvision cudatoolkit=10.2
conda install -c conda-forge opencv hydra-core pycocotools tensorboard faiss 
conda install -c iopath -c fvcore -c conda-forge fvcore
conda install -v -c vissl -c fvcore -c iopath -c conda-forge -c pytorch -c defaults vissl=0.1.5
```

There are some background things to be aware of, but none are the primary problem here:

- The recommended way to install faiss is with `faiss-gpu` from the `pytorch` channel, but we depend on `faiss` which is in conda-forge. They have different dependencies, but the conda-forge one depends on pytorch and we need to avoid pulling in a `pytorch` from a default channel, which would be an unofficial build. Hence installing pytorch before any conda-forge packages. Ideally we would switch to depend on faiss-gpu. After that change, I would hope that the first two `conda install` lines could be swapped.
- There is a tensorboard package on both the `anaconda` and `conda-forge` channels and they have different dependencies.
- I have an initial build of vissl (0.1.0) sitting on anaconda cloud. This is from before we decided to put the classy vision code inside the vissl package instead of its own package. I specify the version of vissl in the above to avoid touching the old one. When the next release happens we should delete all the previous ones.
- The three packages `scikit-learn`, `scipy`, and `parameterized` exist both in `anaconda` and `conda-forge` and it doesn't seem to matter which ones you use.

When I try the same recipe with python 3.8, it chokes at the conda-forge stage. Just `conda install -c conda-forge opencv`  alone is problematic for conda. The opencv package for Python 3.8 on conda-forge seems to be quite unfriendly to install. If you start an environment with `conda create -p /raid/reizenstein/jeremyconda/foo python=3.8` and immediately try `conda install -c conda-forge opencv`, it works but insists on a downgrade from Python 3.8.8 to Python 3.8.3. If I try `conda install -c defaults -c conda-forge opencv` it works more smoothly. There may be a way forward here to make things work.

Basically opencv is a bit "special". It would be better to not explicitly depend on opencv, and tell the user to install it via pip. Similar to how our pip packages work. Or even, could we use PIL for the functionality we currently use in opencv, and therefore not need it at all.

**Comment by prigoyal** [comment]
thank you @bottler ! It seems like our immediate fix is to just recommend `python=3.7` in the conda installation to the user. This is probably something I should update the website and our install instructions immediately.

For our next release, I agree with all the suggestions. I think opencv is a dependency but we can make it optional as we only have 0.1% of the code that requires opencv and otherwise we use PIL.
For the faiss, we should probably just switch to faiss-gpu but again , this is also a dependency that can become optional. I'll add a follow-up for both of these tasks!

**Comment by pclucas14** [comment]
Hi, 

I tried the following commands with python 3.7

```
 conda create -n vissl2 python=3.7
conda activate vissl2
conda install -c pytorch pytorch=1.7.1 torchvision cudatoolkit=10.2
conda install -c vissl -c iopath -c conda-forge -c pytorch -c defaults apex vissl
```

And still I'm getting 

```
(vissl2) lucaspc@devfair0224:~/repos/swav$ conda install -c vissl -c iopath -c conda-forge -c pytorch -c defaults apex vissl
Solving environment: failed
```

**Comment by prigoyal** [comment]
thank you @pclucas14 , we are working on building new packages with the dependencies reduced! In the meantime, if you are blocked, I recommend 2 options:
1. If conda is must, build vissl from scratch following the instructions https://github.com/facebookresearch/vissl/blob/master/INSTALL.md#Install-from-source-in-Conda-environment 
2. Use PIP packages of VISSL instead

**Comment by bottler** [comment]
I think the InvalidVersionSpecError comes from using an old version of conda itself. I am using the 2020.11 anaconda module on devfair, which is conda 4.9.2, and which I think is the recommended one.

**Comment by pclucas14** [comment]
ok willdo, thanks Priya!

**Comment by pclucas14** [comment]
OK let me try with a better conda version then

**Comment by pclucas14** [comment]
@bottler can you show me what your devfair outputs for `conda info` ?

I'm using the conda located in `base environment : /public/apps/anaconda3/5.0.1  (read only)` which is version 4.5.8, and `conda update conda` fails

**Comment by prigoyal** [comment]
@pclucas14 , just an update here: @bottler uploaded new v0.1.5 conda packages that should now also work with python=3.8. 
You should still have the conda 4.9.2 or latest conda otherwise all good to go! :)

**Comment by JinYeJin** [comment]
I updated conda version with just `conda update conda`, and it worked for me.
